File: engineering/codebase/smartqr.frontend-services/src/domain/codes/common/models/CodeDto.ts
import type { Temporal } from "temporal-polyfill";

import type { BarcodeFormat, ContentMode, ContentType } from "../../content";
import type { CodeRuleDto } from "../../rules";
import type { CodeStyleDto } from "../../style";

/** Represents an issued code — its identity, wire settings, style, and the rules carrying its content. */
export interface CodeDto {
  /** The code's unique id. */
  id: string;

  /** The short slug printed on a dynamic code; absent on a static code. */
  slug?: string;

  /** The short URL a dynamic code encodes; absent on a static code. */
  shortUrl?: string;

  /** The code's display name. */
  name: string;

  /** The symbology the code renders as. */
  barcodeFormat: BarcodeFormat;

  /** How the symbol resolves — baked payload (static) or short link (dynamic). Fixed at create. */
  mode: ContentMode;

  /** The kind of content every rule of this code carries. */
  contentType: ContentType;

  /** Whether the code is active. */
  isActive: boolean;

  /** The lifetime scan count; always zero on a static code, which never reaches the server. */
  scanCount: number;

  /** When the code was created. */
  createdAt: Temporal.Instant;

  /** The routing rules, each carrying the content it serves. */
  rules: CodeRuleDto[];

  /** The persisted visual style. */
  style: CodeStyleDto;
}

File: engineering/codebase/smartqr.backend-services/SmartQr.Application/Codes/Core/Models/CodeDto.cs
using WoW.Two.Sdk.Backend.Beta.Codes.Models.Style;
using SmartQr.Common.Domain.Codes.Core.Enums;
using SmartQr.Domain.Codes.Core.Enums;
using SmartQr.Domain.Codes.Rules.Models;

namespace SmartQr.Application.Codes.Core.Models;

/// <summary>Represents a code in API shape.</summary>
public sealed record CodeDto
{
    /// <summary>Gets the id of the code.</summary>
    public required Guid Id { get; init; }

    /// <summary>Gets the public slug of a dynamic code; null on a static code.</summary>
    public string? Slug { get; init; }

    /// <summary>Gets the short URL a dynamic code resolves through; null on a static code.</summary>
    public string? ShortUrl { get; init; }

    /// <summary>Gets the display name of the code.</summary>
    public required string Name { get; init; }

    /// <summary>Gets the rendering symbology of the code.</summary>
    public required BarcodeFormat BarcodeFormat { get; init; }

    /// <summary>Gets how the code's symbol resolves.</summary>
    public required ContentMode Mode { get; init; }

    /// <summary>Gets the kind of content every rule of the code carries.</summary>
    public required CodeContentType ContentType { get; init; }

    /// <summary>Gets whether the code currently resolves.</summary>
    public bool IsActive { get; init; }

    /// <summary>Gets the running scan total of the code.</summary>
    public long ScanCount { get; init; }

    /// <summary>Gets the timestamp when the code was created.</summary>
    public DateTimeOffset CreatedAt { get; init; }

    /// <summary>Gets the routing rules, each carrying the content it serves.</summary>
    public IReadOnlyList<CodeRuleValueObject> Rules { get; init; } = [];

    /// <summary>Gets the persisted visual style, or the render default when none was saved.</summary>
    public required StyleSpec Style { get; init; }
}
